File: java/src/ik/Graph.java
package ik;

import java.util.*;
import java.io.InputStreamReader;
import java.io.BufferedReader;

/**
 * This class is the heart of the project to determine intrinsically
 * knotted graphs.  It is objects of this Graph class that represents the
 * graphs that we are testing.  The classification tests are performed on
 * these objects.
 */
public class Graph {
  private String      name;
  private int         order;
  private boolean[][] data;

  /**
   * Creates a new Graph object.
   *
   * @param name A String name for the graph
   * @param order An integer number of vertices for the graph
   */
  public Graph(String name, int order) {
    this.name  = name;
    this.order = order;
    this.data  = new boolean[order][order];
  }

  /**
   * The name of the graph.
   *
   * @return A String name for the graph.
   */
  public String getName() {
    return name;
  }

  /**
   * Returns the number of vertices in the graph.
   *
   * @return An integer representing the number of vertices in the graph.
   */
  public int getOrder() {
    return order;
  }

  /**
   * Returns the number of edges in the graph.
   *
   * @return An integer representing the number of edges in the graph.
   */
  public int getSize() {
    return edges().size();
  }

  /**
   * Adds the given edge to the Graph.
   *
   * @param fromVertex The from vertex for the edge
   * @param toVertex The to vertex for the edge
   */
  public void addEdge(int fromVertex, int toVertex) {
    setEdge(fromVertex, toVertex, true);
  }

  /**
   * Determines whether or not this graph includes the provided edge.
   *
   * @param fromVertex The from vertex for the edge
   * @param toVertex The to vertex for the edge
   * @return A boolean value indicating whether or not this edge is a part
   *         of the graph.
   */
  public boolean hasEdge(int fromVertex, int toVertex) {
    return getEdge(fromVertex, toVertex);
  }

  /**
   *  The list of edges in this graph.
   *
   * @return A List of edges that are a part of this graph.
   */
  public List<int[]> edges() {
    List<int[]> edges = new ArrayList<int[]>(maxSize());

    for (int from = 0; from < getOrder(); from++) {
      for (int to = from + 1; to < getOrder(); to++) {
        if (hasEdge(from, to)) {
          int[] edge = {from, to};
          edges.add(edge);
        }
      }
    }

    return edges;
  }

  /**
   * Returns a list of vertices that are connected via an edge to the
   * provided vertex.
   *
   * @param fromVertex The vertex we are investigating.
   * @return A List of integers which represent edges from the fromVertex
   *         provided to the vertices returned.
   */
  public List<Integer> connectedVertices(int fromVertex) {
    List<Integer> connectedVertices = new ArrayList<Integer>(getOrder());

    for (int toVertex = 0; toVertex < getOrder(); toVertex++) {
      if (fromVertex != toVertex && hasEdge(fromVertex, toVertex)) {
        connectedVertices.add(new Integer(toVertex));
      }
    }

    return connectedVertices;
  }

  /**
   * Create a new graph with the supplied edge contracted.
   *
   * @param fromVertex The from vertex for the edge.
   * @param toVertex The to vertex for the edge.
   * @return A new Graph with the provided edge removed.
   */
  public Graph contractEdge(int fromVertex, int toVertex) {
    Graph minor = null;

    // Cannot contract an edge if we don't have one
    if (hasEdge(fromVertex, toVertex)) {
      minor = new Graph(getName(), getOrder() - 1);
      int      maxVertex    = fromVertex > toVertex ? fromVertex : toVertex;
      int      minVertex    = fromVertex > toVertex ? toVertex   : fromVertex;
      Iterator edgeIterator = edges().iterator();

      while (edgeIterator.hasNext()) {
        int[] edge    = (int[])edgeIterator.next();
        int   newFrom = -1;
        int   newTo   = -1;

        // The vertex labels for the edge may have changed as a result of
        // one less vertex
        if (edge[0] == maxVertex) {
          newFrom = minVertex;
        } else if (edge[0] > maxVertex) {
          newFrom = edge[0] - 1;
        } else {
          newFrom = edge[0];
        }

        if (edge[1] == maxVertex) {
          newTo = minVertex;
        } else if (edge[1] > maxVertex) {
          newTo = edge[1] - 1;
        } else {
          newTo = edge[1];
        }

        if (newFrom != newTo) {
          minor.addEdge(newFrom, newTo);
        }
      }
    }

    return minor;
  }

  /**
   * Create a new graph with the vertices provided removed.
   *
   * @param vertices An array of vertices to remove from our graph
   * @return A new Graph without the vertices provided
   */
  public Graph removeVertices(int[] vertices) {
    Graph subGraph = new Graph(getName(), getOrder() - vertices.length);

    Iterator edgeIterator = edges().iterator();

    while (edgeIterator.hasNext()) {
      int[] edge = (int[])edgeIterator.next();
      int   from = edge[0];
      int   to   = edge[1];

      // If either of the points is in the vertices array, then the point
      // is thrown out
      boolean disregardEdge = false;
      for (int i = 0; i < vertices.length; i++) {
        if (from == vertices[i] || to == vertices[i]) {
          disregardEdge = true;
        }
      }

      if (!disregardEdge) {
        int origFrom = from;
        int origTo   = to;

        // The vertex labels for the edge may have changed as a result of
        // less vertices
        for (int i = 0; i < vertices.length; i++) {
          if (origFrom > vertices[i]) {
            from--;
          }
          if (origTo > vertices[i]) {
            to--;
          }
        }

        subGraph.addEdge(from, to);
      }
    }

    return subGraph;
  }

  /**
   * Determine if our graph contains as a minor the provided graph.
   *
   * @param minor The Graph we are comparing to our graph.
   * @return A boolean value indicating whether or not our graph
   *         contains as a minor the provided graph.
   */
  public boolean containsMinor(Graph minor) {
    if (getOrder() < minor.getOrder()) { return false; }
    if (getSize()  < minor.getSize())  { return false; }
    if (containsSubgraph(minor))       { return true;  }

    // Contracting an edge will result in one less edge and one less vertex
    // We check to see if we can afford to lose one edge and one vertex
    if (getSize() <= minor.getSize() || getOrder() <= minor.getOrder()) {
      return false;
    }

    Iterator edgeIterator = edges().iterator();

    // Try contracting each edge, then check again
    while(edgeIterator.hasNext()) {
      int[] edge = (int[])edgeIterator.next();

      Graph newMinor = contractEdge(edge[0], edge[1]);

      if (newMinor.containsMinor(minor)) {
        return true;
      }
    }

    return false;
  }

  /**
   * Determine if our graph contains the provided graph as a subgraph.
   *
   * @param subGraph The graph we are testing against our own.
   * @return A boolean value indicating whether or not our graph
   *         contains the provided graph as a subgraph.
   */
  public boolean containsSubgraph(Graph subGraph) {
    boolean containsSubgraph = false;

    // Has to have at least as many edges and vertices as the subgraph
    // in order to even be considered
    if (getSize()  >= subGraph.getSize() &&
        getOrder() >= subGraph.getOrder()) {
      // Initialize an empty vertexMap from our graph to the subgraph
      int[] vertexMap = new int[subGraph.order];
      for (int i = 0; i < vertexMap.length; i++) {
        vertexMap[i] = -1;
      }

      // Do the work
      containsSubgraph = searchForSubgraph(subGraph, vertexMap, 0);
    }

    return containsSubgraph;
  }

  /**
   * Determines if two graphs are isomorphic to each other.
   *
   * @param graph The graph to check to see if we are isomorphic to.
   * @return A boolean value indicating whether or not we are isomorphic.
   */
  public boolean isIsomorphicTo(Graph graph) {
    return containsSubgraph(graph)       &&
           getSize()  == graph.getSize() &&
           getOrder() == graph.getOrder();
  }

  /**
   * Does the recursive work of trying to determine if a mapping from our
   * graph to the subgraph can be found.
   *
   * @param subGraph The Graph that we are trying to determine if we are a
   *                 subgraph of
   * @param vertexMap The current mapping from our vertices to those of the
   *                  subgraph.
   * @param nextIndex The index we will be working with next.
   * @return A boolean value indicating if the mapping 'so far' is a complete
   *         mapping to the subgraph.
   */
  private boolean searchForSubgraph(Graph subGraph,
                                    int[] vertexMap,
                                    int   nextIndex) {
    // We first need to verify our current map state and make sure that it
    // is a 'possible' solution for what we know so far
    if (!verifyPartialMapping(subGraph, vertexMap, nextIndex - 1)) {
      return false;
    }

    // If we are trying to work on an index that is past our length of the
    // the vertexMap, that means that we have a valid mapping
    if (vertexMap.length == nextIndex) {
      return true;
    }

    // Walk through each vertex
    for (int currentIndex = 0; currentIndex < this.order; currentIndex++) {
      if (!containsVertex(vertexMap, currentIndex)) {
        // Assume this vertex is correct, and then test it
        vertexMap[nextIndex] = currentIndex;
        if (searchForSubgraph(subGraph, vertexMap, nextIndex+1)) {
          return true;
        }
        else {
          // It wasn't correct, try the next
          vertexMap[nextIndex] = -1;
        }
      }
    }

    return false;
  }

  /**
   * Checks whether the current mapping that has been constructed for the
   * graph to a possible subgraph 'is possible'.  It just confirms the
   * in progress solution.
   *
   * @param subGraph The Graph that we are trying to determine if we are a
   *                 subgraph of
   * @param vertexMap The current mapping from our vertices to those of the
   *                  subgraph.
   * @param workingIndex The index we are working with currently.
   * @return A boolean value indicating if the mapping 'so far' is
   *         possible or not.
   */
  private boolean verifyPartialMapping(Graph subGraph,
                                       int[] vertexMap,
                                       int   workingIndex) {
    boolean successfulMapping = true;

    if (workingIndex >= 0 && workingIndex < vertexMap.length) {
      int      mappedFromVertex  = vertexMap[workingIndex];
      List     connectedVertices = subGraph.connectedVertices(workingIndex);
      Iterator verticesIterator  = connectedVertices.iterator();

      while (verticesIterator.hasNext() && successfulMapping) {
        int toVertex       = ((Integer)verticesIterator.next()).intValue();
        int mappedToVertex = vertexMap[toVertex];
        if (mappedToVertex >= 0) {
	      // If the edge exists in the subgraph but not in our graph via
	      // the proposed mapping, then we are done and this mapping doesn't
	      // work
          successfulMapping = this.hasEdge(mappedFromVertex, mappedToVertex);
        }
      }
    }

    return successfulMapping;
  }

  /**
   * Determines if the vertexMap contains the vertex provided.  It is
   * merely a lookup into the array for the value.
   *
   * @param vertexMap The current array of integer vertices.
   * @param vertex    The vertex to check.
   * @return A boolean indicating whether or not the vertex was found.
   */
  private boolean containsVertex(int[] vertexMap, int vertex) {
    boolean containsVertex = false;
    int     currentIndex   = 0;

    while ((currentIndex < vertexMap.length) && (!containsVertex)) {
      containsVertex = (vertexMap[currentIndex] == vertex);
      currentIndex++;
    }

    return containsVertex;
  }

  /**
   * Sets the value for an edge.
   *
   * @param fromVertex The starting vertex of the edge.
   * @param toVertex   The ending vertex of the edge.
   * @param value      The boolean value to set it to, where true indicates
   *                   the edge exists, false indicates that it does not.
   */
  private void setEdge(int fromVertex, int toVertex, boolean value) {
    validateEdge(fromVertex, toVertex);

    data[fromVertex][toVertex] = data[toVertex][fromVertex] = value;
  }

  /**
   * Returns the value set for the edge.
   *
   * @return A boolean result where true means there was an edge, and false
   *         means there was not.
   */
  private boolean getEdge(int fromVertex, int toVertex) {
    validateEdge(fromVertex, toVertex);

    return data[fromVertex][toVertex];
  }

  /**
   * Validates that the edge is legitimate.
   *
   * @throws IllegalArgumentException if it is invalid.
   */
  private void validateEdge(int fromVertex, int toVertex) {
    validateVertex(fromVertex);
    validateVertex(toVertex);

    if (fromVertex == toVertex)
      throw new IllegalArgumentException("Loops are not allowed: " +
                                         fromVertex+" == "+toVertex);
  }

  /**
   * Validates that the vertex is legitimate.
   *
   * @throws IllegalArgumentException if it is invalid.
   */
  private void validateVertex(int vertex) {
    if (vertex < 0 || vertex >= getOrder())
      throw new IllegalArgumentException("Vertex is not in the valid range:" +
                                         " 0 <= "+vertex+" < "+getOrder());
  }

  /**
   * Determines the maximum number of edges that are possible with the given
   * order.  This is used to know how bit to create Arrays.
   *
   * @return An integer representing the max possible size.
   */
  private int maxSize() {
    return (getOrder() * (getOrder() - 1)) / 2;
  }
}

File: ruby/lib/graph.rb
# This class is the heart of the project to determine intrinsically
# knotted graphs.  It is objects of this Graph class that represents the
# graphs that we are testing.  The classification tests are performed on
# these objects.  The 'meaty' parts of this class are the minor_of?, 
# contains_minor?, subgraph_of?, and isomorphic_to? methods.
#
# An object of type Graph is immutable.  This means that upon initialization
# the edges of the graph are set and cannot be modified.  Any calls to 
# remove_vertex, remove_edges, contract_edge will result in a new graph
# being created.  The current graph will not be manipulated.
#
# This class does assume that the nauty tools--dretog, labelg, and planarg
# are in the environment PATH.
class Graph
  attr_reader :name
  attr_reader :order
  attr_reader :edges 
  attr_reader :safe_mode
  
  # Creates a new graph.
  #
  # The name is an arbitrary name for the graph.
  # The order is the number of vertices in the graph.
  # The edges are an array of all of the edges in the graph.
  # The safe mode allows checks to be performed on inputs to the graph.  If
  # set, edges and vertices will be confirmed to be valid before proceeding.
  # The reason that this is optional is because these validations,
  # while valuable, result in roughly a 30% performance hit during the
  # classification efforts.  Since over 99% of the graphs created in this
  # project are created indirectly as a result of a call to 'minor_of' we know
  # that the graphs are valid and do not need any validation, and thus it 
  # is safe to remove these tests.  If this class were used 'outside' of this 
  # single approach, then the conditions would not necessarily always be met.
  # For this reason the 'safe_mode' option is available and can be set to
  # which will perform the validations.
  #
  # Edges are simply two element arrays (ex. [1,2]).
  #
  # We create an adjaceny matrix in order to make 'easy' operations much 
  # more effecient.
  def initialize(name, order, edges, safe_mode=true)
    @name       = name
    @order      = order
    @edges      = edges
    @safe_mode  = safe_mode
    @adj_matrix = (0...@order).to_a.collect{Array.new(@order)}
    
    if @safe_mode
      Graph.validate_edges!(edges, @order)
      @edges = Graph.canonicalize_edges(@edges)
    end
    
    @edges.each do |edge|
      @adj_matrix[edge[0]][edge[1]] = @adj_matrix[edge[1]][edge[0]] = true
    end
  end
  
  # The number of edges in the graph.
  def size
    edges.size
  end
  
  # Determines if the supplied edge exists in the graph or not.
  def has_edge?(edge)
    Graph.validate_edge!(edge, order) if @safe_mode
    
    @adj_matrix[edge[0]][edge[1]]
  end
  
  # The number of edges attached to the supplied vertex.
  def degree(vertex)
    Graph.validate_vertex!(vertex, order) if @safe_mode
    
    histogram[vertex]
  end
  
  # An array of the number of edges attached to every vertex from 0 to order.
  def histogram
    @histogram ||= (0...order).to_a.collect{|vert| @adj_matrix[vert].nitems}
  end
  
  # Removes the vertex from the graph, returning a new graph without that 
  # vertex and the edges connected to that vertex.
  def remove_vertex(vertex)
    Graph.validate_vertex!(vertex, order) if @safe_mode
    
    new_order = order - 1
    new_edges = []
    
    edges.each do |from, to|
      next if from == vertex || to == vertex
      
      new_from = from > vertex ? from - 1 : from
      new_to   = to   > vertex ? to   - 1 : to
      
      new_edges << [new_from, new_to]
    end
    
    Graph.new(name, 
              new_order, 
              new_edges, 
              false)
  end
  
  # Removes all of the vertices from the graph, returning a new graph
  # without those vertices and edges attached to those vertices.
  def remove_vertices(*vertices)
    # Note: This can be optimized to remove multiple vertices in one go,
    #       instead of removing one vertex at a time
    vertices.sort.reverse.inject(self) do |new_graph, vertex| 
      new_graph.remove_vertex(vertex)
    end
  end
  
  # Removes all of the edges from the graph, returning a new graph
  # without those edges.
  def remove_edges(*edges_to_remove)
    if @safe_mode
      unless edges_to_remove.all?{|edge| has_edge?(edge)}
        raise "Cannot remove an edge that does not exist."
      end
      
      edges_to_remove = Graph.canonicalize_edges(edges_to_remove)
    end
    
    Graph.new(name,
              order,
              edges - edges_to_remove,
              false)
  end
  
  # Contracts the given edge, returning a new graph with that edge contracted.
  # Contracting an edge [0,1] means to bring the vertex 1 on to 0, so edges
  # that were previously connected to 1 are now connected to 0, and the
  # vertex 1 no longer exists.
  def contract_edge(edge)
    if @safe_mode
      unless has_edge?(edge)
        raise "Cannot contract an edge which does not exist."
      end
      
      edge = Graph.canonicalized_edge(edge)
    end
    
    new_order = order - 1
    new_edges = []
    
    edges.each do |from, to|
      new_from = if from == edge[1]
                   edge[0]
                 elsif from > edge[1]
                   from - 1
                 else
                   from
                 end

      new_to   = if to == edge[1]
                   edge[0]
                 elsif to > edge[1]
                   to - 1
                 else
                   to
                 end

      # Do not allow loops
      new_edges << [new_from, new_to] if new_from != new_to
    end
    
    Graph.new(name, 
              new_order, 
              Graph.canonicalize_edges(new_edges), 
              false)
  end

  # Determines if two graphs, no matter the vertex labelling, are identical.
  def isomorphic_to?(graph)
    return false unless order          == graph.order
    return false unless size           == graph.size
    return false unless histogram.sort == graph.histogram.sort
    
    to_canonical == graph.to_canonical
  end
  
  # Determines if the our graph contains as a minor, the supplied graph.
  # Optionally can allow two graphs that are isomorphic to be considered 
  # minors.
  def contains_minor?(graph, check_isomorphic=true)
    graph.minor_of?(self, check_isomorphic)
  end
  
  # Determines if our graph is a minor of the supplied graph.  Optionally
  # can allow two graphs that are isomorphic to be considered minors.
  def minor_of?(graph, check_isomorphic=true)
    return false if order > graph.order
    return false if size  > graph.size
    return true  if subgraph_of?(graph, check_isomorphic)
    
    # Contracting an edge will result in one less vertex and at least
    # one less edge
    if order < graph.order && size < graph.size
      graph.edges.each do |edge|
        new_graph = graph.contract_edge(edge)
        
        return true if minor_of?(new_graph)
      end
    end
    
    false
  end 

  # Determines if our graph is a subgraph of the supplied graph.  Optionally
  # can allow two graphs that are isomorphic to be considered subgraphs.
  def subgraph_of?(graph, check_isomorphic=true)
    return false if order > graph.order
    return false if size  > graph.size
    
    if order < graph.order
      # Get the number of vertices equivalent in both graphs, then check
      remove_count = graph.order - order
      
      each_combination(graph.order, remove_count) do |vertices_to_remove|
        # Note: Could be optimized by determining how many edges would be
        #       removed by removing these vertices, and checking that this
        #       number is at least as many as edges in self.size
        new_graph = graph.remove_vertices(*vertices_to_remove)
        
        return true if subgraph_of?(new_graph)
      end
    elsif size < graph.size
      # Get the number of edges equivalent in both graphs, then check
      remove_count = graph.size - size
      
      each_combination(graph.size, remove_count) do |edge_indexes_to_remove|
        edges_to_remove = graph.edges.values_at(*edge_indexes_to_remove)
        new_graph       = graph.remove_edges(*edges_to_remove)
        
        return true if subgraph_of?(new_graph)
      end
    elsif check_isomorphic
      return isomorphic_to?(graph)
    end
    
    false
  end
  
  # Determines if the given graph is planar.  Planar means that the graph 
  # could be arranged in a plane with no edges crossing over each other.
  def planar? 
    # Note: I imagine there is more pruning that could be added in here if
    #       it is necessary.  This hadn't proved to be particularly slow.
    return false if order >= 3 && size > (3 * order - 6)
    
    results = `echo \"#{to_dreadnaut}\" | dretog -q | planarg -qu 2>&1`
        
    return false if results =~ /0 graphs planar$/
    return true  if results =~ /1 graphs planar$/
    
    raise "ERROR: Unable to process planarg results '#{results}'."
  end
  
  # The dreadnaut form looks like this for the complete graph on 7 vertices:
  # n=7g1 2 3 4 5 6;2 3 4 5 6;3 4 5 6;4 5 6;5 6; 6.
  def to_dreadnaut
    return @dreadnaut if @dreadnaut

    buckets = Array.new(@order)
    edges.each do |edge|
      (buckets[edge[0]] ||= []) << edge[1]
    end
    
    edge_list = buckets.collect{|bucket| bucket.join(' ') if bucket}.join(';')
    
    @dreadnaut = "n=#{order}g#{edge_list}."
  end
  
  # This is a universal, simplistic labeling for a graph that no matter
  # the labelling of vertices, two graphs that are isomorphic will have the
  # same canonical label.
  def to_canonical
    @canonical ||= `echo \"#{to_dreadnaut}\" | dretog -q | labelg -q`
  end 
  
  # This is the format that can be interpreted by the graph_parser.
  def to_s
    "Graph #{name}, order #{order}.\n" +
    "#{order} #{size}\n"               +
    "#{edges.collect{|edge| edge.join(' ')}.join('  ')}\n"
  end
  
  # The 'opposite' of a graph.  Where there is an edge in a graph, its 
  # complement does not have an edge.  Likewise, where there is no edge in
  # a graph, the complement does have an edge.
  def complement
    Graph.new("#{name}_complement", 
              order, 
              Graph.complete_graph(order).edges - edges, 
              false)
  end
  
  # Creates a graph with all of the possible edges for a graph of the 
  # given order.
  def self.complete_graph(order)
    edges = []
    
    order.times do |from|
      ((from + 1)...order).each do |to|
        edges << [from, to]
      end
    end
    
    Graph.new("K#{order}", order, edges, false)    
  end

  # Utility methods
  
  # Make sure that the edges are uniq, sorted, and each edge is in the form
  # [0,1] and not [1,0].
  def self.canonicalize_edges(edges)
    edges.collect{|edge| Graph.canonicalized_edge(edge)}.uniq.sort
  end
  
  # Makes an edge in the form [0,1] and not [1,0]
  def self.canonicalized_edge(edge)
    edge[0] <= edge[1] ? edge : edge.reverse
  end
  
  # Confirms that for the given order, that all of the edges are valid.
  def self.validate_edges!(edges, order)
    raise "Edges cannot be nil." unless edges
    
    edges.each{|edge| Graph.validate_edge!(edge, order)}
  end

  # Confirms that for the given order the edge is valid.
  def self.validate_edge!(edge, order)
    Graph.validate_vertex!(edge[0], order)
    Graph.validate_vertex!(edge[1], order)

    if edge[0] == edge[1]      
      raise "Loops are not allowed: #{edge[0]} != #{edge[1]}"
    end
  end
  
  # Confirms that for the given order the vertex is valid.
  def self.validate_vertex!(vertex, order)
    unless vertex >= 0 && vertex < order
      raise "Vertex is not in the valid range: 0 <= #{vertex} < #{order}"
    end
  end
     
  private
    # Utility method to execute a block for each possible combination of 
    # selection_size elements chosen from total_size elements.  The block
    # will be passed the elements (0 based) in the current combination.
    def each_combination(total_size, selection_size, &block)
      set        = Array.new(selection_size)
      next_index = 0
      last_value = -1

      while next_index >= 0 && next_index < selection_size
        if last_value < (total_size - 1)
          set[next_index]  = last_value + 1
          last_value      += 1
          next_index      += 1
        else
          last_value  = set[next_index - 1]
          next_index -= 1
        end

        if next_index == selection_size
          yield set
          last_value  = set[next_index - 1]
          next_index -= 1
        end
      end
    end
end
